### ISSUE TITLE
No warning for partly evaluated functions at toplevel

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6236]
**Original bug ID:** 6236
**Reporter:** gmelquiond
**Status:** acknowledged (set by @damiendoligez on 2013-11-30T22:28:26Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 4.00.1
**Category:** typing

## Bug description

let f ~a b = failwith b
let _ = f "success"

The compiler emits no warning about the last line not being executed completely. The bug looks obvious on this testcase, but it is mind-boggling on a larger piece of code spanning several files. (We had it occur in Why3 recently.)

So it would be great if warning 5 ("this function application is partial, maybe some arguments are missing") was also emitted for anonymous let-statements at top level. (I know that, if "let ()" had been used instead, the compiler would have complained loudly.)

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Intentionally, let _ = expr does not cause any warning.
Using it is at top-level is bad practice.
You should either write
   let () = ...
or
   let () = ignore (...)
which will both warn you.

**Comment by vicuna** [comment]
**Comment author:** @alainfrisch

What about a new warning on all "let _ = ..." bindings?  In my opinion, even for local bindings, it is bad style.

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

Please let us not multiply coding-style warnings that could easily be implemented by an external tool. In this case, even emacs should be able to warn you.

**Comment by yallop** [comment]
This has been fixed by #1974, released in OCaml 4.08:

```
$ cat partial.ml 
let f ~a b = failwith b
let _ = f "success"
$ ocamlc partial.ml 
File "partial.ml", line 2, characters 8-19:
2 | let _ = f "success"
            ^^^^^^^^^^^
Warning 5: this function application is partial,
maybe some arguments are missing.
```
